community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java

```diff
@@ -228,22 +230,21 @@ private void flushAll()
     /**
      * Insert the given batch of updates into the index defined by the given {@link IndexDescriptor}.
      *
-     * @param descriptor the index descriptor.
+     * @param population the index population.
      * @param batch the list of updates to insert.
      */
-    private void flush( IndexDescriptor descriptor, List<NodePropertyUpdate> batch )
+    private void flush( IndexPopulation population, List<NodePropertyUpdate> batch )
     {
         activeTasks.incrementAndGet();
 
         executor.execute( () -> {
             try
             {
-                IndexPopulation population = populations.get( descriptor );
                 population.populator.add( batch );
             }
-            catch ( Throwable t )
+            catch ( Throwable failure )
             {
-                fail( descriptor, t );
+                populationFailures.add( new PopulationFailure( population.descriptor, failure ) );
             }
             finally
             {
```

Comment:
This is basically how it is right now - population is failed right in the `finally` block. It is a simple and nice solution but than data structure where populations are stored should be synchronized. Currently it is a `ConcurrentHashMap` and it is quite expensive to iterate it for each node.